### PR TITLE
feature: Refactored Router and Controller system. Fixes #26

### LINKED ISSUE #26: Proposal: Controller and Router refactor
I post a proposal for Router and Controller refactoring.
Here are the different parts of a request processing chain. We have to take care to keep the whole chain asynchronous (this is not the case with the actual code!).

The first part is routing request to actual handling code:
I propose to keep the already implemented action anotations because it leave Controllers loosely coupled with the Router. It also keep the responsibility to the router to handle controller instantiation/release. The router must also handle a (configurable) timeout before auto closing a request with an error (to avoid bad async code going nowhere)

The second part is the actual Controller: This part is public and must be implemented by the end user in his application. Because the code must be async, I suggest to instantiate one controller per request to avoid accidental property sharing. The controller must expose to the end-user properties like request/responce. It must also provide a helpers methods to populate request. Finaly, it must also provide a method to auto send (terminate) the responce to the client.

Lastly the responce must be handled by our application. For now we will just handle basic requests like json or Raw content data. The code must be extensible to others content (like html from a template renderer).

My intent is to implement two sorts of Controllers: HttpControllers that handle http requests, and ConsoleControllers that handle console requests (imagine something like tsfw generate database?). ConsoleControllers are out of the question for now but I think the naming is pretty good even if we never implement these.

Okay so here is my proposal:
- [x] Keep the actual route registering (using anotation)
- [x] Keep the actual router but:
      - [x] Instantiate a new controller on each requests
      - [x] populate request/responce/send variable with correct objects in each slot
      - [x] Dispatch it asynchronously
- [x] Keep the actual Controller but:
      - [x] Rename ControllerResult to HttpController (because it contains methods to handle http things)
      - [x] Make HttpController a subclass of Controller
      - [ ] Make all demo Controllers in test app a subclass of HttpController (because, well, they are Http Controller)

I'm actually working on a PR to show this implementation working. However feel free to discuss and add your opinion here.

### PR DESCRIPTION
**Pull request by atrakeur**
See issue #26 for more info.

### CODE REVIEW

File: src/ts-framework-tests/app/controllers/ShopController.ts [hunk 1]
```diff
@@ -1,19 +1,23 @@
...
+import {HttpController} from "../../../ts-framework/Controller/HttpController";
 
-export class ShopController extends Controller
+export class ShopController extends HttpController
 {
     // GET: /shop
     @action() index()
     {
-        this.response.sendContent("Hello");
+        this.content("Hello");
```
Comment (jhonmike) [inline src/ts-framework-tests/app/controllers/ShopController.ts]: suggestion: `new ContentResult()` It seems more pleasant, besides leaving cleaner HttpController

File: src/ts-framework-tests/app/controllers/ShopController.ts [hunk 2]
```diff
@@ -1,19 +1,23 @@
...
     // GET: /shop/product/:id
-    @action() product(id: number)
+    @action({
+        'path': '/shop/product/:id',
+        'method': ['GET']
+    }) product()
     {
-        return new JsonResult({
-            "id": id || 0
+        this.json({
```
Comment (jhonmike) [inline src/ts-framework-tests/app/controllers/ShopController.ts]: suggestion: `new JsonResult()`

File: src/ts-framework-tests/app/controllers/ShopController.ts [hunk 3]
```diff
@@ -1,19 +1,23 @@
...
+            "id": this.request.params().id || 0
```
Comment (jhonmike) [inline src/ts-framework-tests/app/controllers/ShopController.ts]: :+1:

File: src/ts-framework-tests/app/controllers/ShopController.ts [hunk 4]
```diff
@@ -1,19 +1,23 @@
...
     @action() index()
     {
-        this.response.sendContent("Hello");
+        this.content("Hello");
     }
 
     // GET: /shop/product/:id
-    @action() product(id: number)
+    @action({
+        'path': '/shop/product/:id',
```
Comment (jhonmike) [inline src/ts-framework-tests/app/controllers/ShopController.ts]: :+1:

File: src/ts-framework/Controller/HttpController.ts [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    protected redirect(url: string, status: number = 302) {
+        this.send(new RedirectResult(url, status));
+    }
+
+    /**
+     * Send string content
+     * @param {string} text
+     * @param {string?} contentType
+    */
+    protected content(text: string, contentType?: string) {
```
Comment (jhonmike) [inline src/ts-framework/Controller/HttpController.ts]: :-1: my opinion, we can discuss

File: src/ts-framework/Controller/HttpController.ts [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        this.send(new ContentResult(text, contentType));
+    }
+
+    /**
+     * Send json content     
+     * @param {} data
+    */
+    protected json(data: {}) {
```
Comment (jhonmike) [inline src/ts-framework/Controller/HttpController.ts]: :-1: my opinion, we can discuss

File: src/ts-framework-tests/app/controllers/IndexController.ts [hunk 7]
```diff
@@ -1,21 +1,20 @@
...
 import {ContentResult} from "../../../ts-framework/View";
 
-export class IndexController extends Controller
+export class IndexController extends HttpController
 {
     // GET: /
     @action() index()
     {
-        return new ContentResult(`
+        return this.content(`
```
Comment (jhonmike) [inline src/ts-framework-tests/app/controllers/IndexController.ts]: suggestion: `new ContentResult(text, contentType)`

File: src/ts-framework-tests/app/controllers/AuthController.ts [hunk 8]
```diff
@@ -19,15 +20,15 @@ export class AuthController extends Controller
...
-            this.response.redirect("/error");
+            this.redirect("/error");
         }
     }
 
     // POST: /auth/logout
     @action() logout(token: string)
     {
-        this.response.sendContent("logout");
+        this.content("logout");
```
Comment (jhonmike) [inline src/ts-framework-tests/app/controllers/AuthController.ts]: suggestion: `new ContentResult(text, contentType)`

File: src/ts-framework-tests/app/controllers/AuthController.ts [hunk 9]
```diff
@@ -1,15 +1,16 @@
...
+export class AuthController extends HttpController
 {
     // GET: /auth
     @action({
         'path': '/auth',
         'method': ['GET']
     }) index()
     {
-        this.response.sendContent("Auth");
+        return this.content("Auth");
```
Comment (jhonmike) [inline src/ts-framework-tests/app/controllers/AuthController.ts]: suggestion: `new ContentResult(text, contentType)`

File: src/ts-framework-tests/app/controllers/ShopController.ts (hunk 1 above)
Comment (atrakeur) [inline src/ts-framework-tests/app/controllers/ShopController.ts]: You also have to pass the send function pointer to all these structures to keep the controller async.
So the code will be new ContentResult("hello", type, this.send). It's add just too much noise where a single and clean helper function can do it. 

Also on a side note: these are helpers functions, not the only way for the end user to do it. If the end user want to write the more verbose version like you, then he still can do it anyway.

Same comment apply for nexts remarks

Comment (jhonmike) [inline src/ts-framework-tests/app/controllers/ShopController.ts]: valid :+1:

**Pull request merged**